game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/standard/DefensiveStandardTest.java

```diff
@@ -0,0 +1,89 @@
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitFirstStrike;
+import static games.strategy.triplea.delegate.battle.steps.MockGameData.givenGameData;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyCollection;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle;
+import java.util.List;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class DefensiveStandardTest {
+
+  @Mock ExecutionStack executionStack;
+  @Mock IDelegateBridge delegateBridge;
+  @Mock BattleActions battleActions;
+
+  @Nested
+  class GetNames {
+    @Test
+    void hasNamesIfStandardUnitAvailable() {
+      final GameData gameData =
+          givenGameData().withDefendingSuicideAndMunitionUnitsDoNotFire(false).build();
+      final BattleState battleState =
+          givenBattleStateBuilder()
+              .defendingUnits(List.of(givenAnyUnit()))
+              .gameData(gameData)
+              .build();
+      final DefensiveStandard defensiveStandard = new DefensiveStandard(battleState, battleActions);
+      assertThat(defensiveStandard.getNames(), hasSize(2));
+    }
+
+    @Test
+    void hasNoNamesIfStandardUnitIsNotAvailable() {
+      final GameData gameData =
+          givenGameData().withDefendingSuicideAndMunitionUnitsDoNotFire(false).build();
+      final BattleState battleState =
+          givenBattleStateBuilder()
+              .defendingUnits(List.of(givenUnitFirstStrike()))
+              .gameData(gameData)
+              .build();
+      final DefensiveStandard defensiveStandard = new DefensiveStandard(battleState, battleActions);
+      assertThat(defensiveStandard.getNames(), hasSize(0));
+    }
+  }
+
+  @Nested
+  class FireAa {
+    @Test
+    void onlyFireStandardUnits() {
+      final GameData gameData =
+          givenGameData().withDefendingSuicideAndMunitionUnitsDoNotFire(false).build();
+      final DefensiveStandard defensiveStandard =
+          new DefensiveStandard(
+              givenBattleStateBuilder().defendingAa(List.of()).gameData(gameData).build(),
+              battleActions);
+
+      defensiveStandard.execute(executionStack, delegateBridge);
+
+      verify(battleActions)
+          .findTargetGroupsAndFire(
+              eq(MustFightBattle.ReturnFire.ALL),
```

Comment:
Any thoughts if we can have something return this value to us?
Perhaps we can create a module that calculates the next step and returns that as a value. IN such a case if we inject that we can then mock and control it, in which case the `findTarggsGroupsAndFire` parameters is not terribly interesting since it mostly comes from mocks, but ultimately testable. In turn, that module can be tested directly with a very functional style and few mocks and particularly no checking that a void method call was called with the right arguments. Thoughts?

Reply:
I'm not sure what you are asking.

Reply:
I'll be direct:
- mocks are testing code smell
- verifying calls to void methods on mocks is an even bigger smell (you're verifying that a side-effect has occurred in very coupled manner).

To simplify and illustrate what we have:
```
StrategyObject mockStrategy = mock(StrategyObject.class);
ClassUnderTest sut = new ClassUnderTest(mockStrategy);

sut.voidCall("some arg");

verify(mockStrategy).sideEffect(eq("some computation"));
```

The above I'd go as far to say is a testing anti-pattern:
- highly coupled to implementation. Change the prod code, even if you just change API you use, you break the test. Not because the feature is broken, but because the implementation is changed.
- it bakes in the void method calls, which are side-effects into the code deeper and tests that we have side-effects invoked
- if you ever mock an object, one must in turn that mocked object with all of the expected parameters to assert it actually returns the stubbed responses. Otherwise who is to say that the actual object behaves that way, in which case the test is not testing anything. For example, if I mock a DB layer to return 5 results from a select, I need to take the real DB and assert that given that select I get 5 results. Otherwise if the real DB throws a syntax error or otherwise, and all I have done is mocked it, then I'll have a bug that the test papered-over and was not sufficient enough to test (so either use a real object or create another test to verify the behavior that was mocked in the orignal test).

To fix the above, one can often transform the object under test to be a function and then have the caller pass the result to the void method call. So in production code, our original example is this:

```
new ClassUnderTest(new StrategyObject()).voidCall("some arg");
```

Instead you can transform that into:

```
var result = ClassUnderTest().computeSomething("some arg");
new StrategyObject().sideEffect(result);
```

Now, you can test `ClassUnderTest` without mocks and you can still test `StrategyObject` just as you would have before.

If things are *really* coupled then a half step that helps is:

```

ClassUnderTest(new ArgTransformFunction(), new StrategyObject()  ).voidCall("some arg");

// where ArgTransformFunction = "some arg" -> "some computation"
``` 

With that latter the test for ClassUnderTest is completely with mocks. Which is good as all the variation is under your control and you can still test "new ArgTransformFunction()" and "new StrategyObject()" independently. In these cases you are roughly able to remove the verify invocations.

Now, how difficult it would be to do such a transformation is a question. The arg conversion strategy can often be injected as a '@VisibleForTest' constructor and this way you move the interesting code under test to its own unit that you can test in isolation without mocks.

---------------


Beyond the above, something I was thinking about, I'd recommend building some harness or characterization tests. I suspect tests that would help most in the restructuring would have setups like "given a sub, and a v2 transport", assert that the executable stack has "first-strike attack and defender return fire".

Reply:
Understood.  I'll work on this once I've got all the battle steps converted.

Reply:
Updating the unit test to be nicer, I'm happy letting things evolve first. In the meantime, is there a reasonable way in which this test could fail? What type of programming errors are possible for that to happen? Have you considered deleting this test?

My 2 cents, such characterization tests are probably going to keep these updates much safer for you. You'll have a lot more confidence that you have not introduced any bugs and because you'll have test coverage you won't have to add unit tests at every step of the way. It's always risky doing the latter approach as it's hard to know if you are unit testing buggy code and thereby asserting defective behavior.

Reply:
Phrased another way, what exactly are we testing here? Is this a valuable test? Will it be deleted the first time we try to any kind of refactoring and is a failure more likely to indicate that the code changed rather than is wrong?

Reply:
For example, there is another test with this:
```
      verify(battleActions)
          .findTargetGroupsAndFire(
              eq(MustFightBattle.ReturnFire.ALL),
              any(),
              eq(false),
              any(),
              any(),
              anyCollection(),
              anyCollection(),
              anyCollection(),
              anyCollection());
    }
```

If both of these fail, one would be left wondering "why was this a valid assertion in the first place?" Why was this test correct, and what was it testing? DisplayName on test and comments in the assertion and in verifications woud help with that.

Tests are great, but a double edged sword as they can become slow and a point of maintenance. When a test fails, the first questions are: "what was this testing?" and "why was this test valid?". If those questions take a lot of time to determine, and it's just one of a dozen tests to fix, then there is substantial cost to change code. Hence, it's very important that tests are overly verbose in what they are testing and why we expect them to be correct and pass. This way someone trying to fix the test can easily know if the test is still valid and what to look for if the test fails

Reply:
@trevan I'm still thinking about this example, I'm not necessarily convinced one way or another if it should remain or go.
The fact there are any matchers for most of the items we are testing is quite good to indicate which values we are really testing. A comment next to the 'false' perhaps could be most of what we need to make it clear why we expect that 'false' value to be there. Maybe ditto with the 'ReturnFire.ALL' enum.

Reply:
@DanVanAtta , I've just deleted the tests.  The really important part that needs to be tested is the matcher but I can't figure out how to match that.  I'll be refactoring the findTargetGroupsAndFire at some point in the future and then the tests will be better.